### ISSUE TITLE
Mingw compiler includes needs enhancement to handle the path ```data/headers/windows/``` correctly 

### ISSUE THREAD

**Issue by araout42** [rapid7/metasploit-framework#17588]
## Steps to reproduce

To test this, one need to have Mingw compiler installed in the $PATH

1. Create a ruby script file  such as 
```
require 'metasploit/framework/compiler/windows'

class MetasploitModule < Msf::Evasion

  def initialize(info={})
    super(merge_info(info,
      'Name'        => 'test',
      'Targets'     => [ ['Microsoft Windows', {}] ],
      'Platform'    => 'win',
      'Arch'        => ARCH_X86,
    ))
  end
  

  def c_template
    @c_template ||= %Q|
#include <windows.h>
#include <rc4.h>


int main(int argc, char **argv)
{
  char str[] = "Hello World!";
  printf("%s\\n", str);
  return 0;
}|
  end

  def run
    m = Metasploit::Framework::Compiler::Mingw::X86.new({ show_compile_cmd: true, f_name: "OK", compile_options: " -lpsapi -lwininet " })
    bin = m.compile_c(c_template)
    print_status("#{bin}")
  end
end
```
at path ```modules/evasion/windows/test.rb```
3. run msfconsole  ```./msfconsole```
4. ```use evasion/windows/test.rb```
5. ```set payload windows/meterpreter/reverse_tcp```
6. ```set lhost 0.0.0.0```
7. ```exploit```

===-> should return mingw output
```
i686-w64-mingw32-gcc OK.c -I /home/araout/metasploit-framework/data/headers/windows/c_payload_util -o OK.exe -ffunction-sections -fno-asynchronous-unwind-tables -fno-ident -O2  -lpsapi -lwininet -Wl,--no-seh

[*] OK.c:3:10: fatal error: rc4.h: Aucun fichier ou dossier de ce type
    3 | #include <rc4.h>
      |          ^~~~~~~
compilation terminated.
```

## Expected behavior
The expected behavior would be to have the ```<rc4.h>``` header file correctly included from the path ```data/headers/windows/```

## Current behavior
The mingw compiler cannot find the header file ```<rc4.h>```

### Metasploit version
Lastest

## Remediation
The code of the mingw compiler set the wrong include folder 
the file ```lib/metasploit/framework/compiler/mingw.rb```

in the lines 
```
INCLUDE_DIR = File.join(Msf::Config.data_directory, 'headers', 'windows', 'c_payload_util')
```

should be replaced by 
```
 INCLUDE_DIR = File.join(Msf::Config.data_directory, 'headers', 'windows')
```

Probably has to change all the includes of the headers in the directory c_payload_util 

for example 

```#include <chacha.h>```
replaced to 
```#include <c_payload_util/chacha.h>```

**Comment by jmartin-r7** [comment]
The current path is correct, the header files in the `data/headers/windows` folder were added before `mingw` compiler support was introduced.  Meaning all header files in that path may not be compatible with `mingw`.  The `c_payload_util` path currently added as the default `INCLUDE_DIR` in `mingw` is however specifically for use with that compiler.

Instead of changing the default include path it may be more valuable to enhance the `Metasploit::Framework::Compiler::Mingw` module to enable extended configuration of the `include` paths or possibly enhance the compile process similar to  `Metasploit::Framework::Compiler::Windows`.

https://github.com/rapid7/metasploit-framework/blob/6d62362b821bc392648ce01cc690e45bb9c7800f/lib/metasploit/framework/compiler/windows.rb#L21-L22

By enabling extension of the include paths we get a more flexible compile process.

**Comment by araout42** [comment]
> The current path is correct, the header files in the `data/headers/windows` folder were added before `mingw` compiler support was introduced. Meaning all header files in that path may not be compatible with `mingw`. The `c_payload_util` path currently added as the default `INCLUDE_DIR` in `mingw` is however specifically for use with that compiler.
> 
> Instead of changing the default include path it may be more valuable to enhance the `Metasploit::Framework::Compiler::Mingw` module to enable extended configuration of the `include` paths or possibly enhance the compile process similar to `Metasploit::Framework::Compiler::Windows`.
> 
> https://github.com/rapid7/metasploit-framework/blob/6d62362b821bc392648ce01cc690e45bb9c7800f/lib/metasploit/framework/compiler/windows.rb#L21-L22
> 
> By enabling extension of the include paths we get a more flexible compile process.


Your answer is very satisfying
I've been working on such an enhancement of the module
It is not finished yet :)
I will pr once i'm done

**Comment by araout42** [comment]
Well after i have made some testings 
the ```Utils.normalize_code``` function parse the ```#includes <XXX>```  and search them in the metsaploit include folder

then if the file is found, it suppress the ```#include <XXXX>``` line from the sources, and replace it by the content of the file included directly 

thus this is a perfect solution for a custom metasploit compiler...

for Mingw-gcc this is not 
Because in my experience it is firmly forbidden to ```redefine reserved identifyer```
especially with different prototype...

for example when including ```stdio.h``` with the ```normalize_code``` function

![image](https://user-images.githubusercontent.com/41875074/221559524-8b9b69f6-e4e3-4387-baae-091378725f0c.png)

### Alternative Solution possibility  
thus i think a good solution would just be to move or make a link for the ```rc4.h ; base64.h ...``` (all the non standard mingw includes) to th directly to the folder ```c_payload_utils ```

i think it could be more relevant

**Comment by araout42** [comment]
> The project supports various platforms meaning symlinks in the repo are discouraged.
> 
> I suspect the approach to my suggestion was taken a bit too narrowly. `Metasploit::Framework::Compiler::Mingw` does not need to adjust the files in the exact same way as `Metasploit::Framework::Compiler::Windows`, if the compiler were to have a default include path set and support for an override or extension of the path set that might meet the needs here.
> 
> To extend on idea and to better restrict scope, once a module provided include path set is injected the compile task setup could be extended to search the include path set to find the header files requested and create a temporary folder where it can copy each requested header file into using the expected structure and then use the new temporary folder as the `INCLUDE_DIR` passed to the compiler execution.

I will try this solution, 
Is there good practices to create temporary directory within metasploit ? as you mentionned the project support various plateforms i'm sure the use of os based command

i will  use ruby's ``` 'tmpdir'   with  Dir.mktmpdir``` 

and make some testings,

Must Note that the includes of standard header files are provided by the compiler,  therefor collecting ```stdio.h``` from the path ```data/headers/windows/``` and include it will not work...

i'm not sure if it is everywhere but the default include path for mingw on my system are 
```/usr/x86_64-w64-mingw32/include/windows.h
/usr/i686-w64-mingw32/include/windows.h```

**Comment by jmartin-r7** [comment]
`Dir.mktmpdir` should be an acceptable pattern.

To allow for standard headers, the search for `include <filename>` values in the paths specified for the compiler class only needs copy files found to the new temp path.  If the pre-processing allows for a header not found early the compile should fallback to search the default compiler provided headers when the compilation occurs.

I would think a consumer of the `Metasploit::Framework::Compiler::Mingw` class would be expected to add header paths only for headers not already provided by the compiler. I believe this is inline with how most compiler processes function as users would only be expected to enhance the include paths locations for headers not already known to the compiler.

**Comment by jmartin-tech** [comment]
Closing as fixed by #17781
